### ISSUE TITLE
composer stores vendor dir in `$XDG_CONFIG_HOME`

### ISSUE THREAD

**Issue by jrobeson** [composer/composer#5041]
I seem to have a vendor directory with all source files in `$XDG_CONFIG_HOME` when they should be stored in `$XDG_DATA_HOME`.  With the XDG specification I would expect to see in:

~/.config/composer/
   composer.json
   (maybe composer.lock)

~/.local/share/composer/
   all source files

or all of the above in `~/.local/share/composer` 

I do have all 3 of XDG_CONFIG_HOME, XDG_DATA_HOME, and XDG_CACHE_HOME set.

**Comment by Seldaek** [comment]
So this is.. only for `composer global require ...` stuff right? I doubt we can or want to change that really, as it'd mean the default for vendor-dir would have to change depending on a bunch of variables, and also it's  not exactly BC :/

You can easily do it yourself by setting the vendor-dir in ~/.config/composer/composer.json though to point to where you want.

**Comment by jrobeson** [comment]
then the only real fix is to ignore `$XDG_CONFIG_HOME` and just drop it all in `$XDG_DATA_HOME`

**Comment by Seldaek** [comment]
It's not impossible to fix I suppose, but it might break people's setup as they might have XDG_DATA_HOME/composer/vendor/bin in their PATH and stuff like that. Unfortunate but not sure if worth fixing.. Or maybe just fix it for new setups? But that can lead to confusion as well.

**Comment by jrobeson** [comment]
you mean `$XDG_CONFIG_HOME/composer/vendor/bin` ? because what you said is the way it should be.

**Comment by jrobeson** [comment]
oh, to clarify, when i said everything under `~/.local/share/composer`, I wasn't suggesting to change the directory structure underneath.

**Comment by Seldaek** [comment]
Yeah I mixed up the name sorry, but anyway you see my point? Right now global composer installs are mainly used for global utilities, so they're in PATH, if we suddenly move it we break stuff.

**Comment by jrobeson** [comment]
yeah, I know it will break stuff. I just wonder for how many, and how this issue went uncovered?

**Comment by alcohol** [comment]
I personally think the whole `global` flag/command is a painful source of confusion either way. I understand it has its use-cases, but in general I think most people have different expectations from it. If anything, it deserves a good refactoring with clear documentation, or just a straight-up removal (cause there are other ways to achieve the same goal).

**Comment by Seldaek** [comment]
@alcohol we know you don't like it, no need to repeat it on every issue related to it :) It's there and IMO it's convenient and useful and there to stay. Not sure what you think is unclear about it? It just needs a PATH entry and then `composer global require/update` lets you install global utils. Sure you can do it without global by using `-d ~/composer` instead of `global` but it doesn't hurt to have, and it's pretty much all it does.

@jrobeson your feedback was taken into account for rollback files AFAIK, but nobody mentioned the vendor dir issue, and I didn't think of it. I think it can be fixed for new setups without doing much breakage, then people can migrate their setup themselves if they care and it'll start using the correct path then.

**Comment by alcohol** [comment]
@seldaek quite a lot of people who are new to composer quickly/often relate it to npm's global. Which is quite different..

**Comment by Seldaek** [comment]
How is it so different? At least I use both for the same.

From NPM docs:

> If you want to use it as a command line tool, something like the grunt CLI, then you want to install it globally. On the other hand, if you want to depend on the package from your own module using something like Node's require, then you want to install locally.

**Comment by alcohol** [comment]
A global install on npm also means you could use it locally though, doesn't it?

**Comment by Seldaek** [comment]
Maybe you could but it sounds like a terrible idea and I have never seen anyone rely on that.

**Comment by alcohol** [comment]
Well would it really break path setups?

Don't you export your path using `PATH="$(composer -g config home)/bin:$PATH"` ?

**Comment by jrobeson** [comment]
@Seldaek : so we want to just put everything in `$XDG_DATA_HOME` ? I don't think we can actually comply to the spec fully, since we can't put composer.lock where it should go either.

**Comment by stof** [comment]
> A global install on npm also means you could use it locally though, doesn't it?

@alcohol not by default. You need an additional env variable to tell Node to look for modules in the global config

@Seldaek IMO, the most confusing part of ``composer global`` is that it installs all global utilities as dependencies of the same global project, which means that it requires all utilities to have compatible deps and it won't use locked dependencies of these tools (meaning things might break, for instance when Satis is not yet updated for a recent change in composer as its lock file still uses the older version)

**Comment by barryvdh** [comment]
Agreed, it would be easy to install magerun/drush/laravel installer etc with `composer global install ..`, but when you have > 1 project, it's not really easy anymore..

**Comment by jcrben** [comment]
Definitely don't think everything should be tossed in the data directory.

Agreed with [this comment](https://github.com/composer/composer/issues/5041#issuecomment-195341741):

````
I think it can be fixed for new setups without doing much breakage, 
then people can migrate their setup themselves if they care and
it'll start using the correct path then.
````

In the meantime having some files in an improper folder isn't the end of the world. Updating the source code doesn't need to remove those files, but users would have to update their PATH going forward.

**Comment by remicollet** [comment]
Another point about XDG installation layout

For now, scripts are installed in vendor/bin
Wouldn't it make sense to have them in .local/bin instead.

The huge benefit is that directory in (usually) in the user PATH.

Again, workaround exists (setting bin-dir or setting PATH), but I think it could be cleaner.

**Comment by imme-emosol** [comment]
So for the time being this might make composer most conforming to XDG standards ?:
```sh
composer global config bin-dir "${XDG_BIN_HOME:-${HOME?}/.local/bin}"
composer global config vendor-dir "${XDG_DATA_HOME:-${HOME?}/.local/share}/composer"
```

---

Got XDG_BIN_HOME from : https://www.freedesktop.org/software/systemd/man/latest/file-hierarchy.html#~/.local/bin/ .
Its not rolled out everywhere, it seems : https://gitlab.freedesktop.org/xdg/xdg-specs/-/issues/14 .

The composer packaged with fedora's has the `Composer\\Factory::useXDG` disabled ( https://src.fedoraproject.org/rpms/composer/blob/rawhide/f/composer-noxdg.patch ).

Adding composer its vendor-bin-dir to PATH ( as mentioned in for instance https://github.com/composer/composer/issues/11439 ) seems to be somewhat contradicting to the idea of standardizing/minimizing desktop paths .

Doing
`export COMPOSER_BIN_DIR="${XDG_BIN_HOME:-${HOME?}/.local/bin}"`
would cause issues (for example https://github.com/composer/composer/issues/6700) with projects outside `${XDG_CONFIG_HOME:-${HOME?}/.config}/composer/` .

Adding a .env ( that has COMPOSER_BIN_DIR ) in the home-dir ( to be parsed by some shell-extension ) seems like too much hackery ( and is less likely to be parsed/used when doing `composer global ..` from some other path ) .

**Comment by TomaszGasior** [comment]
In XDG-enabled systems Composer should by default put global binary files to ~/.local/bin. Many systems like Ubuntu or Fedora by default add ~/.local/bin directory to the $PATH if it exists (Composer may have to create ~/.local/bin if its the first app to add new files there.)

As an example: Python's pip by default adds its user-wide global binaries to ~/.local/bin.

**Comment by Seldaek** [comment]
If someone can provide a PR that is backwards compatible (i.e. changes this for new setup but not if the global vendor (for XDG_DATA_HOME) or vendor/bin (for XDG_BIN_HOME) dir exists), then sure.. I personally probably won't find time to spend on this as we clearly see by the fact it's been 9years.
